source/module_cell/parallel_kpoints.cpp

```diff
@@ -194,104 +194,8 @@ void Parallel_Kpoints::pool_collection(double *valuea, double *valueb, const Mod
     assert( a.getBound3() == b.getBound3() );
     assert( a.getBound4() == b.getBound4() );
     const int dim = dim2 * dim3 * dim4;
-#ifdef __MPI
-    const int ik_now = ik - this->startk_pool[this->my_pool];
-    const int begin = ik_now * dim2 * dim3 * dim4;
-    double* pa = &a.ptr[begin];
-    double* pb = &b.ptr[begin];
-
-    const int pool = this->whichpool[ik];
-
-	GlobalV::ofs_running << "\n ik=" << ik;
-
-    if (this->rank_in_pool==0)
-    {
-        if (this->my_pool==0)
-        {
-            if (pool==0)
-            {
-                for (int i=0; i<dim ; i++)
-                {
-                    valuea[i] = *pa;
-                    valueb[i] = *pb;
-                    ++pa;
-                    ++pb;
-                }
-            }
-            else
-            {
-				GlobalV::ofs_running << " receive data.";
-                MPI_Status ierror;
-                MPI_Recv(valuea, dim, MPI_DOUBLE, this->startpro_pool[pool], ik*2+0, MPI_COMM_WORLD,&ierror);
-                MPI_Recv(valueb, dim, MPI_DOUBLE, this->startpro_pool[pool], ik*2+1, MPI_COMM_WORLD,&ierror);
-            }
-        }
-        else
-        {
-            if (this->my_pool == pool)
-            {
-				GlobalV::ofs_running << " send data.";
-                MPI_Send(pa, dim, MPI_DOUBLE, 0, ik*2+0, MPI_COMM_WORLD);
-                MPI_Send(pb, dim, MPI_DOUBLE, 0, ik*2+1, MPI_COMM_WORLD);
-            }
-        }
-    }
-	else
-	{
-		GlobalV::ofs_running << "\n do nothing.";
-	}
-    MPI_Barrier(MPI_COMM_WORLD);
-
-    /*
-
-
-    	if(this->whichpool[ik] == GlobalV::MY_POOL)
-    	{
-    		if(GlobalV::MY_POOL > 0 && GlobalV::RANK_IN_POOL == 0)
-    		{
-    			// data transfer ends.
-    			MPI_Send(pa, dim, MPI_DOUBLE, 0, ik*2,   MPI_COMM_WORLD);
-    			MPI_Send(pb, dim, MPI_DOUBLE, 0, ik*2+1, MPI_COMM_WORLD);
-    		}
-    		else if(GlobalV::MY_POOL == 0 && GlobalV::MY_RANK == 0)
-    		{
-    //			std::cout << "\n ik = " << ik << std::endl;
-    			// data transfer begin.
-    			for(int i=0; i<dim; i++)
-    			{
-    				valuea[i] = *pa;
-    				valueb[i] = *pb;
-    				++pa;
-    				++pb;
-    			}
-    			// data transfer ends.
-    		}
-    	}
-    	else
-    	{
-    		if(GlobalV::MY_RANK==0)
-    		{
-    			MPI_Status* ierror;
-    			const int iproc = this->startpro_pool[ this->whichpool[ik] ];
-    			MPI_Recv(valuea, dim, MPI_DOUBLE, iproc, ik*2,   MPI_COMM_WORLD,ierror);
-    			MPI_Recv(valueb, dim, MPI_DOUBLE, iproc, ik*2+1, MPI_COMM_WORLD,ierror);
-    		}
-    	}
-    	*/
-#else
-    // data transfer ends.
-    const int begin = ik * dim2 * dim3 * dim4;
-    double* pa = &a.ptr[begin];
-    double* pb = &b.ptr[begin];
-    for (int i=0; i<dim; i++)
-    {
-        valuea[i] = *pa;
-        valueb[i] = *pb;
-        ++pa;
-        ++pb;
-    }
-    // data transfer ends.
-#endif
+    pool_collection_aux(valuea, a, dim, ik);
+    pool_collection_aux(valueb, b, dim, ik);
```

Comment:
I have a comment here. I think the developer who wrote these lines did not appropriately abstract functions here. Because operations involving so-called `a` and `b` are not universal or say "everywhere", naming something with `a` and `b` can only make code look worse.
If you have time, I suggest you change them with more meaningfull name.
Thanks

Reply:
In fact, this overloading of this function hasn't been used anywhere yet, so I can't tell exactly what `a` and `b` mean. But judging from ComplexArray later on and realArray here, it might have something to do with the two ways of storing complex arrays. One where each element of the array is a complex<double>, and one where the real and imaginary parts of the complex numbers are stored in two separate real arrays. 

So I'm considering changing a and b to `re` and `im`.